File: LC/[0813]  Largest Sum of Averages.java
/*
* partition nums into k (non-empty) sub-arrays -> max score
* where score = sum of avg of subarrays
* -> find optimized solution from partition nums into k -> DP IV
*/
class Solution {
    public double largestSumOfAverages(int[] nums, int k) {
        int N = nums.length;
        double[][] dp = new double[N+1][k+1];
        //dp[ii][kk]: max score for nums[:ii+1], k=kk
        //dp[ii][kk] = max(dp[ii][kk], avg(nums[jj:ii+1]+dp[ii-1][jj-1]))
        for(int ii=0; ii<=N; ++ii){
            for(int jj=0; jj<=k; ++jj) dp[ii][jj] = Integer.MIN_VALUE;
        }
        dp[0][0] = 0.0; //nums=None, no partion -> max score=0
        
        //java insert dummy to beginning of int array
        /*
        1. Java convert int array to List of Integers
        2. Java insert element into List
        3. Java convert List of Integer to int array
        */
        var list = IntStream.of(nums).boxed().collect(Collectors.toCollection(ArrayList::new));
        list.add(0, 0);
        //java map list Integer to int
        nums = list.stream().mapToInt(ii->ii).toArray();//java inserted dummy int to beginning of an array

        for(int ii=1; ii<=N; ++ii){//ii=0 is the dummy.  For each element as last element in last partition
            for(int kk=1; kk<=Math.min(k, ii); ++kk){//from 1 partition to k partition.
                double ss = 0;
                for(int jj=ii; jj>=kk; --jj){//jj is the starting index of last partition.
                    ss += nums[jj];
                    dp[ii][kk] = Math.max(dp[ii][kk], ss/(ii-jj+1)+dp[jj-1][kk-1]);
                }
            }

        }

        double ans = 0;
        for(int kk=1; kk<=k; ++kk) ans = Math.max(ans, dp[N][kk]);

        return ans;
    }
}

File: LC/[0813]  Largest Sum of Averages.cpp
class Solution {
public:
    double largestSumOfAverages(vector<int>& nums, int k) {
        int N = nums.size(); //take size
        nums.insert(nums.begin(), 0);//insert dummy 
        auto dp = vector<vector<double>>(N+1, vector<double>(k+1, INT_MIN));

        //for(int ii=0; ii<=N; ++ii)dp[ii][0]=INT_MIN;
        dp[0][0]=0;
        for(int ii=1; ii<=N; ++ii){//for each element (we have dummy header)
            //let kk start from 2 and initialize kk=1 can be another way doing this
            for(int kk=1; kk<=min(k, ii); ++kk){//for each num of partition
                //for(int jj=k; jj<=ii; ++jj){//try each starting index of last partition
                double sum=0;
                for(int jj=ii; jj>=kk; --jj){//jj is the starting index of last subarray
                    sum+=nums[jj];
                    dp[ii][kk] = max(dp[ii][kk], dp[jj-1][kk-1]+sum/(ii-jj+1));//[0:jj] with kk-1 partitions and [jj:N+1] with one partition
                }
            }
        }
        double ans=0;
        for(int kk=1; kk<=k; ++kk){
            ans = max(ans, dp[N][kk]);
        }
        return ans;
    }
};